File: ldplayer-bridge/src/services/instance_record_store.py

"""
Registro persistente POR INSTANCIA: health, apks revisadas/instaladas,
permisos confirmados, últimos eventos (reboot/launch/quit), próximo
check programado, tareas, etc.

Un archivo por instancia en DATA_DIR/instances/<index>.json. A
diferencia de health/<index>.json (que es solo un cache de health con
TTL, se pisa entero en cada refresh) y status/all.json (snapshot
efímero del monitor), este archivo es un REGISTRO acumulativo: se lee,
se modifica parcialmente, y se vuelve a escribir completo.

Lo leen y ESCRIBEN los dos lados:
  - Python (este módulo) registra lo que se ejecuta físicamente sobre
    la instancia: health, apks instaladas, permisos otorgados/revocados,
    reboot/launch/quit, y el próximo check programado del monitor.
  - Node (services/instanceRecordStore.js) registra lo que orquesta:
    los pasos/tareas del pipeline de setup (deviceSetupPipeline.js).

Para que no se pisen si escriben casi al mismo tiempo, cada
actualización hace lock -> leer -> modificar -> escribir -> unlock,
con un lockfile simple (<archivo>.json.lock) creado de forma exclusiva
(O_CREAT|O_EXCL, atómico también en Windows). El equivalente en Node
usa el MISMO nombre de archivo y el mismo protocolo, así que ambos
procesos se coordinan sin necesidad de compartir lenguaje ni proceso.
"""
import json
import os
import random
import time
import uuid
from typing import Any, Callable, Dict, List, Optional

from config import settings

MAX_EVENTS = 100
MAX_TASKS = 100
LOCK_TIMEOUT_S = 5.0
LOCK_STALE_S = 10.0
LOCK_RETRY_MIN_MS = 30
LOCK_RETRY_MAX_MS = 120


class InstanceRecordStore:
    def __init__(self, base_dir: str, owner: str = "python"):
        self.dir = os.path.join(base_dir, "instances")
        self.owner = owner
        os.makedirs(self.dir, exist_ok=True)

    def _path(self, index: int) -> str:
        return os.path.join(self.dir, f"{index}.json")

    def _lock_path(self, index: int) -> str:
        return f"{self._path(index)}.lock"

    # ------------------------------------------------------------------
    # Lock cross-proceso / cross-lenguaje (archivo .lock exclusivo)
    # ------------------------------------------------------------------
    def _acquire_lock(self, index: int) -> None:
        lock_path = self._lock_path(index)
        deadline = time.time() + LOCK_TIMEOUT_S
        while True:
            try:
                fd = os.open(lock_path, os.O_CREAT | os.O_EXCL | os.O_WRONLY)
                with os.fdopen(fd, "w") as f:
                    f.write(f"{self.owner}:{os.getpid()}:{time.time()}")
                return
            except FileExistsError:
                age = None
                try:
                    age = time.time() - os.path.getmtime(lock_path)
                except OSError:
                    pass
                if age is not None and age > LOCK_STALE_S:
                    # lock huérfano: el que lo tomó se cayó antes de soltarlo
                    try:
                        os.remove(lock_path)
                    except OSError:
                        pass
                    continue
                if time.time() > deadline:
                    raise TimeoutError(
                        f"No se pudo tomar el lock de instances/{index}.json "
                        f"en {LOCK_TIMEOUT_S}s (¿el otro proceso quedó trabado?)"
                    )
                time.sleep(random.uniform(LOCK_RETRY_MIN_MS, LOCK_RETRY_MAX_MS) / 1000)

    def _release_lock(self, index: int) -> None:
        try:
            os.remove(self._lock_path(index))
        except OSError:
            pass

    # ------------------------------------------------------------------
    # Lectura / escritura atómica del registro
    # ------------------------------------------------------------------
    def _read_raw(self, index: int) -> Dict[str, Any]:
        path = self._path(index)
        if not os.path.exists(path):
            return self._blank(index)
        try:
            with open(path, "r", encoding="utf-8") as f:
                data = json.load(f)
                return data if isinstance(data, dict) else self._blank(index)
        except (json.JSONDecodeError, OSError):
            return self._blank(index)

    def _write_raw(self, index: int, data: Dict[str, Any]) -> None:
        path = self._path(index)
        tmp_path = f"{path}.tmp-{os.getpid()}-{time.time()}"
        with open(tmp_path, "w", encoding="utf-8") as f:
            json.dump(data, f, ensure_ascii=False, separators=(",", ":"))
        os.replace(tmp_path, path)

    @staticmethod
    def _blank(index: int) -> Dict[str, Any]:
        return {
            "index": index,
            "name": None,
            "updated_at": None,
            "updated_by": None,
            "health": {},
            "schedule": {
                "next_check_at": None,
                "last_reboot_at": None,
                "last_launch_at": None,
                "last_quit_at": None,
            },
            "apks": {},
            "permissions": {},
            "tasks": [],
            "events": [],
        }

    def get(self, index: int) -> Dict[str, Any]:
        return self._read_raw(index)

    def update(self, index: int, updater: Callable[[Dict[str, Any]], None]) -> Dict[str, Any]:
        """Lock -> leer -> updater(record) muta el dict in-place -> escribir -> unlock."""
        self._acquire_lock(index)
        try:
            record = self._read_raw(index)
            updater(record)
            record["updated_at"] = time.time()
            record["updated_by"] = self.owner
            self._write_raw(index, record)
            return record
        finally:
            self._release_lock(index)

    def delete(self, index: int) -> None:
        path = self._path(index)
        if os.path.exists(path):
            try:
                os.remove(path)
            except OSError:
                pass

    # ------------------------------------------------------------------
    # Helpers de alto nivel (todos pasan por update() -> con lock)
    # ------------------------------------------------------------------
    def record_health(self, index: int, name: Optional[str], health: Dict[str, Any]) -> None:
        def _fn(r):
            if name:
                r["name"] = name
            r["health"] = {
                "android_started": health.get("android_started"),
                "pid": health.get("pid"),
                "battery": health.get("battery"),
                "battery_error": health.get("battery_error"),
                "checked_at": time.time(),
            }
        self.update(index, _fn)

    def schedule_next_check(self, index: int, seconds_from_now: float) -> None:
        def _fn(r):
            r["schedule"]["next_check_at"] = time.time() + seconds_from_now
        self.update(index, _fn)

    def record_launch(self, index: int) -> None:
        def _fn(r):
            r["schedule"]["last_launch_at"] = time.time()
        self.update(index, _fn)
        self.add_event(index, "launch", "Instancia lanzada")

    def record_reboot(self, index: int) -> None:
        def _fn(r):
            r["schedule"]["last_reboot_at"] = time.time()
        self.update(index, _fn)
        self.add_event(index, "reboot", "Instancia reiniciada")

    def record_quit(self, index: int) -> None:
        def _fn(r):
            r["schedule"]["last_quit_at"] = time.time()
        self.update(index, _fn)
        self.add_event(index, "quit", "Instancia cerrada")

    def record_apk(self, index: int, apk_id: str, status: str,
                    apk_path: Optional[str] = None) -> None:
        """status: 'installed' | 'uninstalled' | 'force_stopped' | 'data_cleared' | 'failed'"""
        def _fn(r):
            entry = r["apks"].get(apk_id, {})
            entry["status"] = status
            if apk_path:
                entry["apk_path"] = apk_path
            entry[f"{status}_at"] = time.time()
            r["apks"][apk_id] = entry
        self.update(index, _fn)

    def record_permission(self, index: int, package_name: str, permission: str, granted: bool) -> None:
        def _fn(r):
            pkg = r["permissions"].setdefault(package_name, {})
            pkg[permission] = {"granted": granted, "confirmed_at": time.time()}
        self.update(index, _fn)

    def add_event(self, index: int, event_type: str, message: str, extra: Optional[dict] = None) -> None:
        def _fn(r):
            events: List[dict] = r.setdefault("events", [])
            events.append({
                "ts": time.time(),
                "type": event_type,
                "source": self.owner,
                "message": message,
                "extra": extra or {},
            })
            if len(events) > MAX_EVENTS:
                del events[: len(events) - MAX_EVENTS]
        self.update(index, _fn)

    def add_task(self, index: int, task_type: str, detail: Optional[dict] = None) -> str:
        task_id = uuid.uuid4().hex[:12]

        def _fn(r):
            tasks: List[dict] = r.setdefault("tasks", [])
            tasks.append({
                "id": task_id,
                "type": task_type,
                "status": "pending",
                "created_at": time.time(),
                "updated_at": time.time(),
                "detail": detail or {},
            })
            if len(tasks) > MAX_TASKS:
                del tasks[: len(tasks) - MAX_TASKS]
        self.update(index, _fn)
        return task_id

    def update_task(self, index: int, task_id: str, status: str, detail: Optional[dict] = None) -> None:
        def _fn(r):
            for task in r.get("tasks", []):
                if task["id"] == task_id:
                    task["status"] = status
                    task["updated_at"] = time.time()
                    if detail:
                        task["detail"] = {**task.get("detail", {}), **detail}
                    break
        self.update(index, _fn)

    def prune(self, active_indices: set) -> None:
        """Borra registros (y locks sueltos) de índices que ya no existen."""
        try:
            files = os.listdir(self.dir)
        except OSError:
            return
        for name in files:
            stem = name[: -len(".json")] if name.endswith(".json") else (
                name[: -len(".json.lock")] if name.endswith(".json.lock") else None
            )
            if stem is None or not stem.isdigit():
                continue
            if int(stem) not in active_indices:
                try:
                    os.remove(os.path.join(self.dir, name))
                except OSError:
                    pass


instance_record_store = InstanceRecordStore(settings.DATA_DIR, owner="python")

File: ldplayer-node-bridge/src/services/instanceRecordStore.js
'use strict';
/**
 * Registro persistente POR INSTANCIA — espejo Node de
 * services/instance_record_store.py (Python). Mismo archivo
 * (DATA_DIR/instances/<index>.json), mismo protocolo de lock
 * (<archivo>.json.lock, O_CREAT|O_EXCL, stale > 10s se descarta,
 * timeout 5s, retry 30-120ms), para que los dos lados puedan
 * escribir sin pisarse sin necesitar coordinación externa.
 *
 * Responsabilidad de Node: registrar lo que ORQUESTA (pasos/tareas
 * de deviceSetupPipeline.js y jobRunner.js). Health, apks, permisos,
 * launch/reboot/quit y next_check_at los escribe Python — no los
 * tocamos acá para no duplicar/pisar esos campos.
 */
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const config = require('../config');

const MAX_EVENTS = 100;
const MAX_TASKS = 100;
const LOCK_TIMEOUT_S = 5.0;
const LOCK_STALE_S = 10.0;
const LOCK_RETRY_MIN_MS = 30;
const LOCK_RETRY_MAX_MS = 120;

function sleep(ms) { return new Promise((r) => setTimeout(r, ms)); }
function randomInt(min, max) { return Math.floor(Math.random() * (max - min + 1)) + min; }

class InstanceRecordStore {
  constructor(baseDir, owner = 'node') {
    this.dir = path.join(baseDir, 'instances');
    this.owner = owner;
    fs.mkdirSync(this.dir, { recursive: true });
  }

  _path(index) { return path.join(this.dir, `${index}.json`); }
  _lockPath(index) { return `${this._path(index)}.lock`; }

  // ------------------------------------------------------------------
  // Lock cross-proceso / cross-lenguaje (mismo protocolo que Python)
  // ------------------------------------------------------------------
  async _acquireLock(index) {
    const lockPath = this._lockPath(index);
    const deadline = Date.now() + LOCK_TIMEOUT_S * 1000;
    for (; ;) {
      try {
        const fd = fs.openSync(lockPath, 'wx'); // O_CREAT|O_EXCL|O_WRONLY, atómico también en Windows
        fs.writeSync(fd, `${this.owner}:${process.pid}:${Date.now() / 1000}`);
        fs.closeSync(fd);
        return;
      } catch (err) {
        if (err.code !== 'EEXIST') throw err;
        let age = null;
        try {
          age = Date.now() - fs.statSync(lockPath).mtimeMs;
        } catch (_) { /* pudo desaparecer justo ahora */ }
        if (age !== null && age > LOCK_STALE_S * 1000) {
          // lock huérfano: el que lo tomó se cayó antes de soltarlo
          try { fs.unlinkSync(lockPath); } catch (_) { }
          continue;
        }
        if (Date.now() > deadline) {
          throw new Error(
            `No se pudo tomar el lock de instances/${index}.json en ${LOCK_TIMEOUT_S}s (¿el otro proceso quedó trabado?)`
          );
        }
        await sleep(randomInt(LOCK_RETRY_MIN_MS, LOCK_RETRY_MAX_MS));
      }
    }
  }

  _releaseLock(index) {
    try { fs.unlinkSync(this._lockPath(index)); } catch (_) { }
  }

  // ------------------------------------------------------------------
  // Lectura / escritura atómica del registro
  // ------------------------------------------------------------------
  _readRaw(index) {
    const filePath = this._path(index);
    if (!fs.existsSync(filePath)) return this._blank(index);
    try {
      const data = JSON.parse(fs.readFileSync(filePath, 'utf8'));
      return data && typeof data === 'object' ? data : this._blank(index);
    } catch (_) {
      return this._blank(index);
    }
  }

  _writeRaw(index, data) {
    const filePath = this._path(index);
    const tmpPath = `${filePath}.tmp-${process.pid}-${Date.now()}`;
    fs.writeFileSync(tmpPath, JSON.stringify(data));
    fs.renameSync(tmpPath, filePath);
  }

  _blank(index) {
    return {
      index,
      name: null,
      window: null,
      updated_at: null,
      updated_by: null,
      health: {},
      schedule: {
        next_check_at: null,
        last_reboot_at: null,
        last_launch_at: null,
        last_quit_at: null,
      },
      apks: {},
      permissions: {},
      tasks: [],
      events: [],
    };
  }

  get(index) { return this._readRaw(index); }

  /** Lock -> leer -> updater(record) muta in-place -> escribir -> unlock. */
  async update(index, updater) {
    await this._acquireLock(index);
    try {
      const record = this._readRaw(index);
      updater(record);
      record.updated_at = Date.now() / 1000;
      record.updated_by = this.owner;
      this._writeRaw(index, record);
      return record;
    } finally {
      this._releaseLock(index);
    }
  }

  delete(index) {
    const filePath = this._path(index);
    if (fs.existsSync(filePath)) {
      try { fs.unlinkSync(filePath); } catch (_) { }
    }
  }

  // ------------------------------------------------------------------
  // Helpers de alto nivel (todos con lock) — lado Node: tareas/eventos
  // de orquestación (pipeline / jobs). NO tocar health/apks/permisos/
  // schedule acá: eso es responsabilidad de Python.
  // ------------------------------------------------------------------
  async addEvent(index, eventType, message, extra = null) {
    return this.update(index, (r) => {
      const events = r.events || (r.events = []);
      events.push({
        ts: Date.now() / 1000,
        type: eventType,
        source: this.owner,
        message,
        extra: extra || {},
      });
      if (events.length > MAX_EVENTS) events.splice(0, events.length - MAX_EVENTS);
    });
  }

  async recordAgent(index, agentInfo) {
    return this.update(index, (r) => {
      r.agent = { ...agentInfo, updated_at: Date.now() / 1000 };
    });
  }
  async addTask(index, taskType, detail = null) {
    const taskId = crypto.randomUUID().replace(/-/g, '').slice(0, 12);
    await this.update(index, (r) => {
      const tasks = r.tasks || (r.tasks = []);
      tasks.push({
        id: taskId,
        type: taskType,
        status: 'pending',
        created_at: Date.now() / 1000,
        updated_at: Date.now() / 1000,
        detail: detail || {},
      });
      if (tasks.length > MAX_TASKS) tasks.splice(0, tasks.length - MAX_TASKS);
    });
    return taskId;
  }
  async recordWindow(index, windowInfo) {
    return this.update(index, (r) => {
      r.window = { ...windowInfo, updated_at: Date.now() / 1000 };
    });
  }
  async recordApp(index, packageName, info) {
    return this.update(index, (r) => {
      const apks = r.apks || (r.apks = {});
      apks[packageName] = { ...(apks[packageName] || {}), ...info };
    });
  }
  async updateTask(index, taskId, status, detail = null) {
    return this.update(index, (r) => {
      for (const task of r.tasks || []) {
        if (task.id === taskId) {
          task.status = status;
          task.updated_at = Date.now() / 1000;
          if (detail) task.detail = { ...task.detail, ...detail };
          break;
        }
      }
    });
  }

  /** Solo se llama del lado Python (monitor.py), que tiene la lista
   * autoritativa de instancias vía LDConsole. No lo dupliques acá para
   * no generar una carrera de borrado entre los dos procesos. */
  prune(activeIndices) {
    let files;
    try { files = fs.readdirSync(this.dir); } catch (_) { return; }
    for (const name of files) {
      let stem = null;
      if (name.endsWith('.json.lock')) stem = name.slice(0, -'.json.lock'.length);
      else if (name.endsWith('.json')) stem = name.slice(0, -'.json'.length);
      if (stem === null || !/^\d+$/.test(stem)) continue;
      if (!activeIndices.has(Number(stem))) {
        try { fs.unlinkSync(path.join(this.dir, name)); } catch (_) { }
      }
    }
  }
}

const instanceRecordStore = new InstanceRecordStore(config.dataDir, 'node');
module.exports = { InstanceRecordStore, instanceRecordStore };